### ISSUE TITLE
Investigate and fix cascading delegate rights

### ISSUE THREAD

**Issue by newmanld** [uclibs/scholar_uc#499]
As a contributor (A), I want the sole ability to determine my delegates, so that I do not have delegates who are unknown to me.  

If I assign another contributor (B) to be my delegate,
  and if that contributor (B) assigns another contributor (C) to be their delegate,
      contributor C should not be a delegate for me (A), unless I separately assign C to be my delegate.

**Comment by hortongn** [comment]
I just did a quick test of this in my local environment.  This can get messy and hard to follow, but here goes...

I have three users: one, two, and three
user one is a delegate of user two
user two is a delegate of user three

As user three, I created a work
As user two, I created a work on behalf of user three
As user two, I also create a work on behalf of myself (user two)
As user one, I created a work on behalf of user two

I logged in as user one
1. I was able to edit the work created by user two on their own behalf
2. I was able to edit the work created by user two on user three's behalf
3. I was NOT able to edit the work created by user three
4. I was NOT able to create a work on behalf of user three.  I could only create on behalf of myself and user two.

I think #1, #3, and #4 work as expected.  I'm not sure about #2.  Should my delegate be able to edit works I create on behalf of someone else?  That may be the situation Amy is seeing.  Ted may be able to edit the works that Amy created on someone else's behalf.  

Another thing that may need to be fixed is that when I'm logged in as user one and go to My Works, the list of works includes the work that user two created on behalf of user three and I could edit that work (as stated above in #2).

**Comment by newmanld** [comment]
Condition #2 should not be allowed.  My delegate should not be able to edit works I create as a delegate for someone else.

Additionally, when I remove someone as a delegate, they should no longer be able to edit my works or submit on my behalf.

**Comment by newmanld** [comment]
If however, I am an editor on a work (or a member of a group with edit rights to a work), that gives me rights independent of delegate status, to edit that specific work.

From: Newman, Linda (newmanld)
Sent: Friday, September 18, 2015 10:47 AM
To: 'uclibs/scholar_uc'; 'uclibs/scholar_uc'
Subject: RE: [scholar_uc] Investigate and fix cascading delegate rights (#499)

Condition #2 should not be allowed.  My delegate should not be able to edit works I create as a delegate for someone else.

Additionally, when I remove someone as a delegate, they should no longer be able to edit my works or submit on my behalf.

**Comment by scherztc** [comment]
I came across all of these issues when I was coding edit access, embargoes, and delegates.  A work has no idea about delegates.  It only know who has access to itself.  So at the moment that you assign a delegate the redis worker checks for all the works that you have edit access to and gives your delegate edit access to these works.  If we want to fix this we need a refactor on roles and rights.

**Comment by newmanld** [comment]
Agreed that a refactor may be needed to prevent unauthorized access.

These prior issues that we have closed are worth looking at, in particular confirming that the worker task to remove rights when a delegate is removed (54) is still functioning.

https://github.com/uclibs/scholar_uc/issues/54
https://github.com/uclibs/scholar_uc/issues/161
https://github.com/uclibs/scholar_uc/issues/250

**Comment by 13miciunasl** [comment]
I used QA for my testing.

User1 creates work, and then makes User2 its delegate. User2 also creates work, then makes User3 its delegate. User3 now has edit access to both User1 and User2's work. User2 now revokes User3's delegate status. Now User3 is denied edit access to User2's work, BUT maintains edit access to all of User1's work.

**Comment by 13miciunasl** [comment]
User1 adds User2 as an editor to a work. User2 makes User3 its delegate, which then causes User3 to be given edit access to User1's work.

**Comment by 13miciunasl** [comment]
There is now only one scenario in which a works edit access will cascade through delegates. First, Bob makes Joe his delegate. Then Joe makes Kevin his delegate. Now if Joe were to create a work on behalf of Bob, the edit access will cascade down and allow Kevin to edit that work (which is owned by Bob, and is NOT desired functionality). All other scenarios regarding cascading work access have been dealt with at this point.
